kubernetes_platform/python/kfp/kubernetes/toleration.py

```diff
@@ -0,0 +1,52 @@
+# Copyright 2023 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List
+
+from google.protobuf import json_format
+from kfp.dsl import PipelineTask
+from kfp.kubernetes import common
+from kfp.kubernetes import kubernetes_executor_config_pb2 as pb
+from kubernetes.client import V1Toleration
+
+
+def add_tolerations(
+    task: PipelineTask,
+    tolerations: List[V1Toleration],
```

Comment:
This is a departure from the existing pattern, where these task methods (/functions) are parameterized by Python primitives, rather than `kubernetes` objects. Is this preferred to keeping with the existing pattern?

WDYT about `add_toleration` (singular), for consistency as well? This makes parameterizing with primitives a bit more manageable.

Reply:
> This is a departure from the existing pattern, where these task methods (/functions) are parameterized by Python primitives, rather than kubernetes objects. Is this preferred to keeping with the existing pattern?

We used a Python Kubernetes client object for a couple of reasons.

First and foremost, that's [what v1 of the SDK does](https://github.com/kubeflow/pipelines/blob/1.8.22/sdk/python/kfp/dsl/_container_op.py#L948), and we wanted to provide as consistent of an experience as possible for users migrating from v1 to v2.

Secondly, node selectors and labels are fairly simple key value pairs, whereas tolerations, with five distinct fields, are substantially more complex -- sufficiently complex to justify using the Python Kubernetes client's `Toleration` object.

> WDYT about add_toleration (singular), for consistency as well? This makes parameterizing with primitives a bit more manageable.

Singular makes more sense. Commit to follow.

Reply:
> First and foremost, that's [what v1 of the SDK does](https://github.com/kubeflow/pipelines/blob/1.8.22/sdk/python/kfp/dsl/_container_op.py#L948), and we wanted to provide as consistent of an experience as possible for users migrating from v1 to v2.

I see the reason for wanting to align with v1. However, it comes at the (possibly greater) cost of making v2 internally consistency. Not going to block on this, but cc @Tomcli for a second opinion.

Reply:
More than happy to refactor. Just let me know. To clarify, the alternative you're referring to would be something along the following lines, right? I could definitely see how this could be more accessible for end users. 

```python
def add_toleration(
    key: Optional[str] = None, 
    operator: Optional[Literal["Equal", "Exists"]] = None,
    value: Optional[str] = None,
    effect: Optional[str] = None,
    toleration_seconds: Optional[int] = None
)
```

Reply:
I would say it comes down to the user preference. The V1Toleration object is useful to standardize and get new updates from the k8s library automatically, but our users often don't understand the k8s object so we have created some helper function package for them to define these objects.

Having `def add_toleration(task: PipelineTask, tolerations: V1Toleration)` makes more sense to me in terms of v1 alignment. Then the add_toleration function could call another function like `def add_toleration_item` with the proto spec key above so that users can have both options.

Just extra information, the reason V1 was using the k8s object is because the SDK compiler was composing the containerSpec in K8S Python Object back then. Since now we are mapping to our own proto file, it make sense to give both options if we care about both backward compatibility and usability https://github.com/kubeflow/pipelines/issues/4454

Reply:
That's a great idea. How about something like this?

```python
def add_toleration_object(
    task: PipelineTask,
    toleration: V1Toleration,
) -> PipelineTask:
    return add_toleration(
        task=task,
        key=toleration.key,
        operator=toleration.operator,
        value=toleration.value,
        effect=toleration.effect,
        toleration_seconds=toleration.toleration_seconds,
    )

def add_toleration(
    task: PipelineTask,
    key: Optional[str] = None,
    operator: Optional[Literal["Equal", "Exist"]] = None,  # https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L3137
    value: Optional[str] = None,
    effect: Optional[Literal["NoSchedule", "PreferNoSchedule", "NoExecute"] = None,  # https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L3146-L3148
    toleration_seconds: Optional[int] = None,
):
    msg = common.get_existing_kubernetes_config_as_message(task)
    msg.tolerations.append(
        pb.Toleration(
            key=key,
            operator=operator,
            value=value,
            effect=effect,
            toleration_seconds=toleration_seconds,
        )
    )
    task.platform_config["kubernetes"] = json_format.MessageToDict(msg)

    return task
```

I think that since we need to support 3.7 we can't use Literals, but we can use Unions or Enums if that's preferable.

Reply:
That sounds good to me. thanks @droctothorpe

Reply:
Thank you, @Tomcli, for the input!

I do want to highlight that since the user code in v2 is syntactically breaking from the user code in v1, the "backward compatibility" with v1 you mention is really only in a stylistic sense, not a true "breaking changes" syntactic sense. We've _already_ departed from v1 conceptually, so using the k8s object in v2 is actually an _additional_ stylistic inconsistency (within v2) on top of the one that already (between v1 and v2). I still tend to think that (1) true backward compatibility (syntax), (2) internal consistency, and (3) backward stylistic consistency should be valued in that order.

That said, I will not block :). Just want to express a difference of opinion to inform the discussion.

In the spirit of a lower maintenance effort and better UX, I recommend we start with just one style, whichever that is. In my experience, multiple ways to do the same thing doesn't make for a better user experience, since the reason for two approaches is typically less obvious to the user than it is to the maintainer(s). Two approaches suggests that the implementations are functionally different in some way, which they are not.

Thank you both for the good discussion and I approve whichever approach you choose.

Reply:
I defer to you on this, Connor, because (a) you're an actual maintainer and I'm just some rando, and (b) your reasoning is unassailable. Just rebased and pushed up an implementation with primitives.

Two quick questions:

1. Is [this](https://github.com/kubeflow/pipelines/pull/10471/files#diff-72e60aa4ca84c37857bc27f55c975ae0f2be7ccabc51a3bf36530ded4332dc55R22-R25) okay? I saw you implemented something similar [in the DSL package](https://github.com/kubeflow/pipelines/blob/6fb997a611118d280325f499491a41799e5948f6/sdk/python/kfp/dsl/__init__.py#L61-L66).

2. The type hints and doc strings point users in the right direction, but they don't enforce operator and effect values. Is that okay? I think we can enforce with [typing.get_args](https://docs.python.org/3/library/typing.html#typing.get_args) if we want to.